src/__tests__/end-to-end.js

```diff
@@ -1,73 +1,150 @@
 import * as React from 'react'
 import {render, waitForElementToBeRemoved, screen, waitFor} from '../'
 
-const fetchAMessage = () =>
-  new Promise(resolve => {
-    // we are using random timeout here to simulate a real-time example
-    // of an async operation calling a callback at a non-deterministic time
-    const randomTimeout = Math.floor(Math.random() * 100)
-    setTimeout(() => {
-      resolve({returnedMessage: 'Hello World'})
-    }, randomTimeout)
-  })
-
-function ComponentWithLoader() {
-  const [state, setState] = React.useState({data: undefined, loading: true})
-  React.useEffect(() => {
-    let cancelled = false
-    fetchAMessage().then(data => {
-      if (!cancelled) {
-        setState({data, loading: false})
-      }
+describe.each([
+  ['real timers', () => jest.useRealTimers()],
+  ['fake legacy timers', () => jest.useFakeTimers('legacy')],
+  ['fake modern timers', () => jest.useFakeTimers('modern')],
+])(
+  'it waits for the data to be loaded in a macrotask using %s',
+  (label, useTimers) => {
+    beforeEach(() => {
+      useTimers()
+    })
+
+    afterEach(() => {
+      jest.useRealTimers()
     })
 
-    return () => {
-      cancelled = true
+    const fetchAMessageInAMacrotask = () =>
+      new Promise(resolve => {
+        // we are using random timeout here to simulate a real-time example
+        // of an async operation calling a callback at a non-deterministic time
+        const randomTimeout = Math.floor(Math.random() * 100)
+        setTimeout(() => {
+          resolve({returnedMessage: 'Hello World'})
+        }, randomTimeout)
+      })
+
+    function ComponentWithMacrotaskLoader() {
+      const [state, setState] = React.useState({data: undefined, loading: true})
+      React.useEffect(() => {
+        let cancelled = false
+        fetchAMessageInAMacrotask().then(data => {
+          if (!cancelled) {
+            setState({data, loading: false})
+          }
+        })
+
+        return () => {
+          cancelled = true
+        }
+      }, [])
+
+      if (state.loading) {
+        return <div>Loading...</div>
+      }
+
+      return (
+        <div data-testid="message">
+          Loaded this message: {state.data.returnedMessage}!
+        </div>
+      )
     }
-  }, [])
 
-  if (state.loading) {
-    return <div>Loading...</div>
-  }
+    test('waitForElementToBeRemoved', async () => {
+      render(<ComponentWithMacrotaskLoader />)
+      const loading = () => screen.getByText('Loading...')
+      await waitForElementToBeRemoved(loading)
+      expect(screen.getByTestId('message')).toHaveTextContent(/Hello World/)
```

Comment:
Do we have a specific reason to use `ByTestId` here instead of `ByText`? Since we advocate against using test id's, I think we can do it too :)

Reply:
> Since we advocate against using test id's, I think we can do it too :)

It doesn't matter what query method we use here since they're not subject of this test.
Our tests are not production code (i.e. actual tests you write in your app). You shouldn't take these as "this is how you're supposed to write tests". I might just use `querySelector` here to make this clear since this test is not about `screen.getByTestId` but `waitForElementremoved`

Reply:
> You shouldn't take these as "this is how you're supposed to write tests".

That's not what I was trying to say, hence the approval.
Since we're using byText in other places in this test, I thought it can work better, but it's your call.